### PR TITLE
Improve performance of String and Vec<u8>

### PR DESCRIPTION
**Pull request by evgenykuzyakov**
Optimizing strings/vec<u8> reads using custom implementation for `Vec<u8>` with a bit of unsafe code, but without UB.
Switching deserialization to use `&mut &[u8]` instead of `BufRead`. It allows us to switch to `BorshError` instead of `std::io::Error` in the future to go with `no_std`.

# Test plan

- Added unit tests for `String` and `Vec<T>`
- Verified gas cost by linking borsh directly to gas fee meter. See https://github.com/nearprotocol/nearcore/issues/1829 
- Re-running benchmarks

### CODE REVIEW

**Comment by evgenykuzyakov** [comment]
> Is it possible that https://doc.rust-lang.org/std/io/struct.BufReader.html implements this very functionality?

I think we can use BufReader with raw fill_buf() functionality for a lot of what I did manually with read_exact

**Comment by evgenykuzyakov** [comment]
## Input Highlights for gas fess:
Cost of Borsh input of (s: String) where s="" is 9,080,640,736
Average diff 131,604,521
Cost of Borsh input of (v: Vec<u8>) where v=[] is 8,455,908,634
Average diff 112,708,303
Cost of Borsh input of (v: Vec<u32>) where v=[] is 8,425,057,666
Average diff 651,364,504

Full results: https://github.com/nearprotocol/nearcore/issues/1829#issuecomment-596137624

TODO: Add Vec<_> tests

**Comment by nearmax** [comment]
Could you rerun https://github.com/nearprotocol/borsh/tree/master/borsh-rs/benchmarks and re-generate the graphs?

File: borsh-rs/borsh/src/de/mod.rs [hunk 1]
```diff
@@ -167,9 +271,67 @@ where
...
+                        len
+                    } else {
+                        result.extend_from_slice(&buf);
+                        buf.len()
+                    }
+                };
+                len -= read_len;
+                reader.consume(read_len)
+            }
+            let result = unsafe {
```
Comment (nearmax) [inline borsh-rs/borsh/src/de/mod.rs]: I don't understand why is it needed. It takes vector `result` and constructs exactly the same vector from its memory.

Comment (evgenykuzyakov) [reply borsh-rs/borsh/src/de/mod.rs]: `result` is `Vec<u8>` while we need `Vec<T>`

We can add an additional check for `size_of::<T>() == size_of::<u8>()`

File: borsh-rs/borsh/src/de/mod.rs [hunk 2]
```diff
@@ -31,11 +31,11 @@ pub trait BorshDeserialize: Sized {
         None
     }
 
-    /// Whether it's possible to transmute `Vec<u8>` directly into `Vec<Self>`.
-    /// If it's `true`, then `Vec<Self>` implementation will use unsafe behavior to transmute
-    /// `Vec<u8>` to `Vec<Self>`. Required `size_of::<Self>() == size_of::<u8>()`.
+    /// Whether Self is u8.
+    /// NOTE: `Vec<u8>` is the most common use-case for serialization and deserialization, it's
+    /// worth handling it as a special case to improve performance.
```
Comment (nearmax) [inline borsh-rs/borsh/src/de/mod.rs]: Please explain that it is to workaround absence of impl specialization and link the Rust issue.

Comment (evgenykuzyakov) [reply borsh-rs/borsh/src/de/mod.rs]: Added

**Comment by nearmax** [comment]
Could you also bump the versions?

File: borsh-rs/borsh/src/de/mod.rs [hunk 3]
```diff
@@ -22,25 +24,98 @@ pub trait BorshDeserialize: Sized {
...
+                ));
+            }
+            buf[0]
+        };
+        reader.consume(size_of::<u8>());
         Ok(res)
     }
+
+    #[inline]
+    fn try_from_slice(v: &[u8]) -> Result<Self, Error> {
```
Comment (nearmax) [inline borsh-rs/borsh/src/de/mod.rs]: We don't really need to override it here. Plus this function has a different behavior than the original that throws when not all bytes were read.

Comment (nearmax) [inline borsh-rs/borsh/src/de/mod.rs]: We don't really need to override it here. Plus this function has a different behavior than the original that throws when not all bytes were read.

File: borsh-rs/borsh/src/de/mod.rs [hunk 4]
```diff
@@ -22,25 +24,98 @@ pub trait BorshDeserialize: Sized {
...
         impl BorshDeserialize for $type {
             #[inline]
-            fn deserialize<R: Read>(reader: &mut R) -> Result<Self, Error> {
-                let mut data = [0u8; size_of::<$type>()];
-                reader.read_exact(&mut data)?;
-                Ok($type::from_le_bytes(data))
+            fn deserialize<R: BufRead>(reader: &mut R) -> Result<Self, Error> {
+                let res = {
+                    let buf = reader.fill_buf()?;
+                    if buf.len() < size_of::<$type>() {
```
Comment (nearmax) [inline borsh-rs/borsh/src/de/mod.rs]: `buf.len()` is the length of the remaining buffer, not the length of the actual input. It can cause several issues:
* First, the buffer length is always under the buffer capacity. By default they use 8KB, but anyone can run borsh deserializer with smaller buffer and then get suprised why this does not work;
* Just because `buf.len() < size_of::<$type>()` does not mean there is not enough data to read from the input. The buffer is populated in discrete steps and if the buffer boundary happens to be passing through the middle of some multi-byte type then we will incorrectly err. I think the reason why we don't catch it with our error messsages is because it is very unlikely though possible.

I suggest instead of trying to workaround this issue, we can use `slice` instead of `BufRead` to avoid thinking of buffering altogether.

Comment (evgenykuzyakov) [reply borsh-rs/borsh/src/de/mod.rs]: The only case when BufRead should return buffer with less capacity than the buffer length is when the capacity of the buffer is less than `size_of::<$type>()`. Of course, it's possible if someone constructs a `BufRead` with capacity less than 8 bytes, but it's not standard scenario. 

> I suggest instead of trying to workaround this issue, we can use `slice` instead of `BufRead` to avoid thinking of buffering altogether.

This is not simple, because we'll have to modify everything else auto-generated that uses `deserialize` right now.

File: borsh-rs/borsh/src/de/mod.rs (hunk 3 above)
Comment (evgenykuzyakov) [reply borsh-rs/borsh/src/de/mod.rs]: Yes, it's not needed for `u8`, but it's needed for other ints and floats.
I think `Vec<u32>` is about 1.5X faster with this vs new `BufRead` with `deserialize`, but I can confirm on gas meter.

Comment (evgenykuzyakov) [reply borsh-rs/borsh/src/de/mod.rs]: Performance difference for `Vec<u32>` is `651,364,504` vs `1,029,288,862` gas per element, so `try_from_slice` has advantage for `u32`

**Comment by evgenykuzyakov** [comment]
Trying switch to `fn deserialize(buf: &mut [u8]) -> Result<Self, Error>;`. 
Will let you know

File: borsh-rs/borsh/src/de/mod.rs (hunk 4 above)
Comment (nearmax) [reply borsh-rs/borsh/src/de/mod.rs]: > The only case when BufRead should return buffer with less capacity than the buffer length

There is probably a typo.  BufRead capacity is always larger or equal to length.

I am not sure I communicated the problem clearly. Let me try again:
* Suppose our BufRead uses 8KB and we are deserializing a large structure (let's say 20KB):
```
struct Large {
field1: F1,
field2: F2,
...
}
```
* Upon first call to `fill_buf` BufRead will read 8KB and set length=8KB and capacity=8KB;
* Fast forward. Suppose we deserialized so far 8KB - 2 bytes, and length=2bytes, while capacity is still 8KB. We are now about to deserialize `field100: u64`;
* We call `u64::deserialize` which first calls `let buf = reader.fill_buf()?;`. But since there are still 2 bytes left it does nothing and returns a slice of 2 bytes, see: https://doc.rust-lang.org/src/std/io/buffered.rs.html#284 Here `self.pos == self.cap - 2`. We then check `buf.len() < size_of::<$type>()` which fails because `buf.len() == 2` and `size_of::<u64>() == 8`.

**Comment by evgenykuzyakov** [comment]
See https://github.com/nearprotocol/borsh/pull/62 for slice reader version. Need to fix benchmarks, but overall looks clean.

**Comment by evgenykuzyakov** [comment]
Going to regenerate benchmarks with the slice run

File: borsh-rs/borsh/src/de/mod.rs (hunk 4 above)
Comment (evgenykuzyakov) [reply borsh-rs/borsh/src/de/mod.rs]: I see. I misunderstood how it works. Good point.

**Comment by evgenykuzyakov** [comment]
Merged buf slice branch, so it's ready to go

File: .gitattributes [hunk 5]
```diff
@@ -1 +1,2 @@
 docs/* linguist-vendored
+docs/criterion/** linguist-generated=true -diff
```
Comment (nearmax) [inline .gitattributes]: How is it different from `linguist-vendored` and what does `-diff` do?

Comment (evgenykuzyakov) [reply .gitattributes]: Also I'm not sure why you would request changes based on the question? Which changes do I need to make?

Comment (evgenykuzyakov) [reply .gitattributes]: https://github.com/github/linguist#vendored-code
https://github.com/github/linguist#generated-code
https://git-scm.com/docs/gitattributes#_marking_files_as_binary

Comment (nearmax) [reply .gitattributes]: I somehow failed to duckduckgo this info. It should've been a comment instead of request for changes.
